### PR TITLE
Document workspaces in The Sway Book

### LINKED ISSUE #1777: Add support for `[workspace]`s to `forc`
## Context

We are progressively coming across more and more cases where it would be great if `forc` could be aware of collections of packages and the way they interact, rather than just reasoning about one package at a time.

**Cargo Workspaces**

Cargo uses the concept of `[workspace]`s to achieve something similar and we may want to draw inspiration from it. However, it's worth noting that cargo's original major motivator was to have the ability for a bunch of related crates to share a compiler output artifact directory (`target/`) to cut down on the epicly long crate build times. Our major motivations are a bit different, and we should keep this in mind when considering the design

### Deploying multiple packages

It is not uncommon for a project to manage a set of contracts that should be deployed simultaneously. Currently, this is a PITA with `forc` as `forc` is only aware of a single package at a time.

It would be nice if `forc` had the ability to reason about a set of packages, and to be able to reason about not just build `[dependencies]`, but **deployment** dependencies and related ordering too. Deployment ordering is important as if contract foo requires interacting with contract bar, then contract foo will likely need contract bar's address which can't be known until it is deployed.

### Testing multiple packages

Currently `forc test` is a thin layer on top of `cargo test` as it only really supports integration tests via the Rust SDK. One common issue with this is that the cargo project root is not always the same as the forc project root, occasionally leading to strange and unexpected behaviour from `forc test`. This is because `cargo` considers the project root to be where the cargo workspace manifest lives, while `forc` considers the project root to be where the forc package manifest lives. This is discussed briefly in #1375 and #1521, and #1612 *might* be related.

If `forc` had the concept of `[workspace]`s, it would at least enable users to setup their project so that both `forc` and `cargo` are aware of the same workspace root, and might make it a bit easier to make `forc test` behaviour more intuitive.

---

cc @adlerjohn finally opening this as a rough tracking issue to get started and somewhere to discuss this. Feel free to add more motivations or details whether in this OP or in new comments!

I remember us having a chat about passing inputs between packages within a hypothetical workspace a while ago in another issue, but can't for the life of me remember where that issue is :thinking: 

Once we have a clearer outline of all the motivations/goals we can think of, I'll have a go at a rough potential design (if no-one beats me to it).

---

# TODO

- [x] #2893
- [x] #2930
- [x] #2931
- [x] #2929
- [ ] Document how workspaces work in the Sway Book.

### LINKED ISSUE #3272: Document workspace support in the Sway Book
After following PRs we will have working workspace building/deploying/checking/updating. We should have some explanation and couple of examples for them:

1. https://github.com/FuelLabs/sway/pull/3134
2. https://github.com/FuelLabs/sway/pull/3146
3. https://github.com/FuelLabs/sway/pull/3248
4. https://github.com/FuelLabs/sway/pull/3255 - closed as already fixed

### PR DESCRIPTION
**Pull request by kayagokalp**
closes #3272.
closes #1777.

blocked by:
- [x] #3248

### CODE REVIEW

File: docs/src/forc/workspaces.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Workspaces
+
+A *workspace* is a collection of one or more packages, namely *workspace members*, that are managed together.
+
+The key points for workspaces are:
+
+* Common `forc` commands available for a single package can also be used for a workspace, like `forc build` or `forc deploy`.
+* All packages share a common [`Forc.lock`] file which resides in the root directory of the workspace.
+
+For workspaces, `Forc.toml` accepts following fields:
```
Comment (mitchmindtree) [inline docs/src/forc/workspaces.md:10]: Workspace manifests are declared within `Forc.toml` files and support the following fields:

File: docs/src/forc/workspaces.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+## The `members` field
+
+The `members` field define which packages are members of the workspace:
+
+```toml
+members = ["member1", "path/to/member2"]
+```
+
+The `members` field accepts entries to be given in relative path with respect to the workspace root.
```
Comment (mitchmindtree) [inline docs/src/forc/workspaces.md:22]:
```suggestion
The `members` field accepts entries to be given in relative path with respect to the workspace root.
Packages that are located within a workspace directory but are *not* contained within the `members` set are ignored.
```

**Review by mitchmindtree — COMMENTED** [review]
Just a few nitpicks, otherwise I think this is a good start! We can flesh out descriptions about build order, deployment behaviour etc in future PRs.

**Review by mitchmindtree — DISMISSED** [review]

**Review by mohammadfawaz — COMMENTED** [review]
Just some nits!

**Review by mohammadfawaz — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**